{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "understood-employer",
   "metadata": {},
   "outputs": [],
   "source": [
    "# take the sample sheet\n",
    "ref_fa=\"/home/hqyone/mnt/2tb/eccDNA/genome/hg38_p13/GRCh38.p13.genome.fa\"\n",
    "ref_gtf=\"/home/hqyone/mnt/2tb/eccDNA/genome/hg38_p13/gtf/gencode.v37.all.sort.gtf.gz\"\n",
    "\n",
    "# project_dir = \"/home/hqyone/mnt/2tb/eccDNA\"\n",
    "\n",
    "# sample_id=\"sample1\"\n",
    "# fastq1=\"{}/GHS21010065/Sample_R21009037-1-D21010704/R21009037-D21010704-D21010704_combined_R1.fastq\".format(project_dir)\n",
    "# fastq2=\"{}/GHS21010065/Sample_R21009037-1-D21010704/R21009037-D21010704-D21010704_combined_R2.fastq\".format(project_dir)\n",
    "# output_dir=\"{}/GHS21010065/Sample_R21009037-1-D21010704\".format(project_dir)\n",
    "\n",
    "# sample_id=\"Cervix_Normal_1\"\n",
    "# fastq1=\"/home/hqyone/mnt/2t/eccDNA/GHS21010065/Sample_R21009038-2-D21010705/R21009038-D21010705-D21010705_combined_R1.fastq\"\n",
    "# fastq2=\"/home/hqyone/mnt/2t/eccDNA/GHS21010065/Sample_R21009038-2-D21010705/R21009038-D21010705-D21010705_combined_R2.fastq\"\n",
    "# output_dir=\"/home/hqyone/mnt/2t/eccDNA/GHS21010065/Sample_R21009038-2-D21010705\"\n",
    "\n",
    "#sample_id=\"Prostate_Cancer_1\"\n",
    "#fastq1=\"/home/hqyone/mnt/2t/eccDNA/GHS21010065/Sample_R21009039-3-D21010706/R21009039-D21010706-D21010706_combined_R1.fastq\"\n",
    "#fastq2=\"/home/hqyone/mnt/2t/eccDNA/GHS21010065/Sample_R21009039-3-D21010706/R21009039-D21010706-D21010706_combined_R2.fastq\"\n",
    "#output_dir=\"/home/hqyone/mnt/2t/eccDNA/GHS21010065/Sample_R21009039-3-D21010706\"\n",
    "\n",
    "# project_dir = \"/home/hqyone/mnt/2tb/eccDNA/data-raw/GHS21030056/Sample_1-1\"\n",
    "# sample_id=\"Sample_1-1\"\n",
    "# fastq1=\"{}/D21040347-D21040347_combined_R1.fastq\".format(project_dir)\n",
    "# fastq2=\"{}/D21040347-D21040347_combined_R2.fastq\".format(project_dir)\n",
    "# output_dir=\"{}/\".format(project_dir)\n",
    "\n",
    "# project_dir = \"/home/hqyone/mnt/2tb/eccDNA/data-raw/GHS21030056/Sample_2-2\"\n",
    "# sample_id=\"Sample_1-2\"\n",
    "# fastq1=\"{}/D21040348-D21040348_combined_R1.fastq\".format(project_dir)\n",
    "# fastq2=\"{}/D21040348-D21040348_combined_R2.fastq\".format(project_dir)\n",
    "# output_dir=\"{}/\".format(project_dir)\n",
    "\n",
    "project_dir = \"/home/hqyone/mnt/2tb/eccDNA/data-raw/GHS21030056/Sample_4-4\"\n",
    "sample_id=\"Sample_4-4\"\n",
    "fastq1=\"{}/D21040350-D21040350_combined_R1.fastq\".format(project_dir)\n",
    "fastq2=\"{}/D21040350-D21040350_combined_R2.fastq\".format(project_dir)\n",
    "output_dir=\"{}/\".format(project_dir)\n",
    "\n",
    "\n",
    "test=False\n",
    "if (test):\n",
    "    test_dir=\"/home/hqyone/mnt/2t/eccDNA/code/test_data\"\n",
    "    output_dir=test_dir\n",
    "    sample_id=\"test\"\n",
    "    test_fastq1=test_dir+\"/test_r1.fq\"\n",
    "    test_fastq2=test_dir+\"/test_r2.fq\"\n",
    "    !head -5000 $fastq1>$test_fastq1\n",
    "    !head -5000 $fastq2>$test_fastq2\n",
    "    fastq1 = test_fastq1\n",
    "    fastq2 = test_fastq2\n",
    "#fastq2=\"\""
   ]
  },
  {
   "source": [
    "# fastqc\n",
    "qc_out_dir=output_dir+\"/qc\"\n",
    "![ ! -d $qc_out_dir ] && mkdir $qc_out_dir\n",
    "!fastqc -f fastq $fastq1 -o $qc_out_dir\n",
    "!fastqc -f fastq $fastq2 -o $qc_out_dir"
   ],
   "cell_type": "code",
   "metadata": {},
   "execution_count": 13,
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "Started analysis of D21040350-D21040350_combined_R1.fastq\n",
      "Approx 5% complete for D21040350-D21040350_combined_R1.fastq\n",
      "Approx 10% complete for D21040350-D21040350_combined_R1.fastq\n",
      "Approx 15% complete for D21040350-D21040350_combined_R1.fastq\n",
      "Approx 20% complete for D21040350-D21040350_combined_R1.fastq\n",
      "Approx 25% complete for D21040350-D21040350_combined_R1.fastq\n",
      "Approx 30% complete for D21040350-D21040350_combined_R1.fastq\n",
      "Approx 35% complete for D21040350-D21040350_combined_R1.fastq\n",
      "Approx 40% complete for D21040350-D21040350_combined_R1.fastq\n",
      "Approx 45% complete for D21040350-D21040350_combined_R1.fastq\n",
      "Approx 50% complete for D21040350-D21040350_combined_R1.fastq\n",
      "Approx 55% complete for D21040350-D21040350_combined_R1.fastq\n",
      "Approx 60% complete for D21040350-D21040350_combined_R1.fastq\n",
      "Approx 65% complete for D21040350-D21040350_combined_R1.fastq\n",
      "Approx 70% complete for D21040350-D21040350_combined_R1.fastq\n",
      "Approx 75% complete for D21040350-D21040350_combined_R1.fastq\n",
      "Approx 80% complete for D21040350-D21040350_combined_R1.fastq\n",
      "Approx 85% complete for D21040350-D21040350_combined_R1.fastq\n",
      "Approx 90% complete for D21040350-D21040350_combined_R1.fastq\n",
      "Approx 95% complete for D21040350-D21040350_combined_R1.fastq\n",
      "Analysis complete for D21040350-D21040350_combined_R1.fastq\n",
      "Started analysis of D21040350-D21040350_combined_R2.fastq\n",
      "Approx 5% complete for D21040350-D21040350_combined_R2.fastq\n",
      "Approx 10% complete for D21040350-D21040350_combined_R2.fastq\n",
      "Approx 15% complete for D21040350-D21040350_combined_R2.fastq\n",
      "Approx 20% complete for D21040350-D21040350_combined_R2.fastq\n",
      "Approx 25% complete for D21040350-D21040350_combined_R2.fastq\n",
      "Approx 30% complete for D21040350-D21040350_combined_R2.fastq\n",
      "Approx 35% complete for D21040350-D21040350_combined_R2.fastq\n",
      "Approx 40% complete for D21040350-D21040350_combined_R2.fastq\n",
      "Approx 45% complete for D21040350-D21040350_combined_R2.fastq\n",
      "Approx 50% complete for D21040350-D21040350_combined_R2.fastq\n",
      "Approx 55% complete for D21040350-D21040350_combined_R2.fastq\n",
      "Approx 60% complete for D21040350-D21040350_combined_R2.fastq\n",
      "Approx 65% complete for D21040350-D21040350_combined_R2.fastq\n",
      "Approx 70% complete for D21040350-D21040350_combined_R2.fastq\n",
      "Approx 75% complete for D21040350-D21040350_combined_R2.fastq\n",
      "Approx 80% complete for D21040350-D21040350_combined_R2.fastq\n",
      "Approx 85% complete for D21040350-D21040350_combined_R2.fastq\n",
      "Approx 90% complete for D21040350-D21040350_combined_R2.fastq\n",
      "Approx 95% complete for D21040350-D21040350_combined_R2.fastq\n",
      "Analysis complete for D21040350-D21040350_combined_R2.fastq\n"
     ]
    }
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "unlike-wisconsin",
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "TrimmomaticPE: Started with arguments:\n",
      " -threads 6 /home/hqyone/mnt/2tb/eccDNA/data-raw/GHS21030056/Sample_4-4/D21040350-D21040350_combined_R1.fastq /home/hqyone/mnt/2tb/eccDNA/data-raw/GHS21030056/Sample_4-4/D21040350-D21040350_combined_R2.fastq /home/hqyone/mnt/2tb/eccDNA/data-raw/GHS21030056/Sample_4-4//for_pair.fq.gz /home/hqyone/mnt/2tb/eccDNA/data-raw/GHS21030056/Sample_4-4//for_unpair.fq.gz /home/hqyone/mnt/2tb/eccDNA/data-raw/GHS21030056/Sample_4-4//rev_pair.fq.gz /home/hqyone/mnt/2tb/eccDNA/data-raw/GHS21030056/Sample_4-4//rev_unpair.fq.gz ILLUMINACLIP:./tools/Trimmomatic-0.39/adapters/TruSeq3-PE-2.fa:2:30:10:2:keepBothReads LEADING:3 TRAILING:3 MINLEN:36\n",
      "Using PrefixPair: 'TACACTCTTTCCCTACACGACGCTCTTCCGATCT' and 'GTGACTGGAGTTCAGACGTGTGCTCTTCCGATCT'\n",
      "Using Long Clipping Sequence: 'AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGTA'\n",
      "Using Long Clipping Sequence: 'AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC'\n",
      "Using Long Clipping Sequence: 'GTGACTGGAGTTCAGACGTGTGCTCTTCCGATCT'\n",
      "Using Long Clipping Sequence: 'TACACTCTTTCCCTACACGACGCTCTTCCGATCT'\n",
      "ILLUMINACLIP: Using 1 prefix pairs, 4 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences\n",
      "Quality encoding detected as phred33\n",
      "Input Read Pairs: 82729305 Both Surviving: 80315476 (97.08%) Forward Only Surviving: 2360064 (2.85%) Reverse Only Surviving: 7844 (0.01%) Dropped: 45921 (0.06%)\n",
      "TrimmomaticPE: Completed successfully\n"
     ]
    }
   ],
   "source": [
    "# trimmomatics\n",
    "# The company Universal Adapter = PrefixPE 1 \n",
    "trimmomatic = \"./tools/Trimmomatic-0.39/trimmomatic-0.39.jar\"\n",
    "output_forward_paired_fq_gz = output_dir+\"/for_pair.fq.gz\"\n",
    "output_forward_unpaired_fq_gz = output_dir+\"/for_unpair.fq.gz\"\n",
    "output_reverse_paired_fq_gz = output_dir+\"/rev_pair.fq.gz\"\n",
    "output_reverse_unpaired_fq_gz = output_dir+\"/rev_unpair.fq.gz\"\n",
    "adapter_file = \"./tools/Trimmomatic-0.39/adapters/TruSeq3-PE-2.fa\"\n",
    "\n",
    "!java -jar $trimmomatic PE -threads 6 $fastq1 $fastq2 $output_forward_paired_fq_gz $output_forward_unpaired_fq_gz $output_reverse_paired_fq_gz $output_reverse_unpaired_fq_gz ILLUMINACLIP:$adapter_file:2:30:10:2:keepBothReads LEADING:3 TRAILING:3 MINLEN:36"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "parental-overhead",
   "metadata": {
    "tags": [
     "outputPrepend"
    ]
   },
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "dev: (893.38, 776.70)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 4383)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 333392 reads in 297.082 CPU sec, 59.297 real sec\n",
      "[M::process] read 333392 sequences (50000071 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (239, 17768, 2205, 226)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (252, 666, 1438)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 3810)\n",
      "[M::mem_pestat] mean and std.dev: (890.79, 864.98)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 4996)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (344, 409, 497)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (38, 803)\n",
      "[M::mem_pestat] mean and std.dev: (382.55, 110.08)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 956)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (344, 713, 1206)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2930)\n",
      "[M::mem_pestat] mean and std.dev: (832.39, 684.75)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3792)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (274, 697, 1719)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4609)\n",
      "[M::mem_pestat] mean and std.dev: (998.08, 1059.05)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6054)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 333390 reads in 298.273 CPU sec, 59.520 real sec\n",
      "[M::process] read 333406 sequences (50000081 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (264, 17756, 2110, 232)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (297, 775, 1691)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4479)\n",
      "[M::mem_pestat] mean and std.dev: (1042.58, 1050.77)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 5873)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (344, 411, 497)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (38, 803)\n",
      "[M::mem_pestat] mean and std.dev: (383.04, 111.18)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 956)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (285, 700, 1171)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2943)\n",
      "[M::mem_pestat] mean and std.dev: (796.35, 656.39)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3829)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (431, 817, 1785)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4493)\n",
      "[M::mem_pestat] mean and std.dev: (1057.25, 924.47)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 5847)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 333392 reads in 299.199 CPU sec, 59.711 real sec\n",
      "[M::process] read 333400 sequences (50000047 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (247, 18234, 2126, 229)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (276, 620, 1364)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 3540)\n",
      "[M::mem_pestat] mean and std.dev: (815.41, 731.66)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 4628)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (345, 411, 503)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (29, 819)\n",
      "[M::mem_pestat] mean and std.dev: (384.82, 111.92)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 977)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (308, 692, 1166)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2882)\n",
      "[M::mem_pestat] mean and std.dev: (788.69, 651.55)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3740)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (277, 762, 1591)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4219)\n",
      "[M::mem_pestat] mean and std.dev: (974.54, 882.39)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 5533)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 333406 reads in 295.314 CPU sec, 58.940 real sec\n",
      "[M::process] read 333402 sequences (50000086 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (269, 17859, 2143, 232)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (246, 581, 1376)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 3636)\n",
      "[M::mem_pestat] mean and std.dev: (783.59, 785.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 4766)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (345, 410, 501)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (33, 813)\n",
      "[M::mem_pestat] mean and std.dev: (384.66, 112.02)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 969)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (282, 700, 1187)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2997)\n",
      "[M::mem_pestat] mean and std.dev: (828.98, 704.24)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3902)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (364, 803, 1615)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4117)\n",
      "[M::mem_pestat] mean and std.dev: (969.21, 868.31)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 5368)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 333400 reads in 301.989 CPU sec, 60.259 real sec\n",
      "[M::process] read 333392 sequences (50000152 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (232, 17600, 2190, 224)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (255, 664, 1393)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 3669)\n",
      "[M::mem_pestat] mean and std.dev: (743.82, 696.26)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 4807)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (347, 410, 498)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (45, 800)\n",
      "[M::mem_pestat] mean and std.dev: (386.56, 106.50)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 951)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (319, 684, 1132)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2758)\n",
      "[M::mem_pestat] mean and std.dev: (770.36, 623.34)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3571)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (344, 868, 1710)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4442)\n",
      "[M::mem_pestat] mean and std.dev: (1000.59, 931.46)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 5808)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 333402 reads in 288.205 CPU sec, 57.509 real sec\n",
      "[M::process] read 333410 sequences (50000105 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (267, 17659, 2262, 203)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (404, 757, 1379)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 3329)\n",
      "[M::mem_pestat] mean and std.dev: (811.41, 617.60)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 4304)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (346, 412, 502)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (34, 814)\n",
      "[M::mem_pestat] mean and std.dev: (384.94, 110.72)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 970)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (341, 709, 1206)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2936)\n",
      "[M::mem_pestat] mean and std.dev: (826.46, 674.50)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3801)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (239, 590, 1342)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 3548)\n",
      "[M::mem_pestat] mean and std.dev: (792.21, 738.10)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 4651)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 333392 reads in 298.553 CPU sec, 59.598 real sec\n",
      "[M::process] read 333400 sequences (50000048 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (230, 17684, 2155, 271)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (284, 665, 1389)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 3599)\n",
      "[M::mem_pestat] mean and std.dev: (852.83, 790.40)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 4704)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (347, 409, 499)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (43, 803)\n",
      "[M::mem_pestat] mean and std.dev: (386.84, 107.63)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 955)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (265, 666, 1106)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2788)\n",
      "[M::mem_pestat] mean and std.dev: (768.05, 648.78)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3629)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (312, 749, 1448)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 3720)\n",
      "[M::mem_pestat] mean and std.dev: (856.71, 757.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 4856)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 333410 reads in 291.484 CPU sec, 58.167 real sec\n",
      "[M::process] read 333404 sequences (50000133 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (246, 17907, 2131, 270)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (314, 675, 1453)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 3731)\n",
      "[M::mem_pestat] mean and std.dev: (893.46, 772.35)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 4870)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (347, 411, 497)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (47, 797)\n",
      "[M::mem_pestat] mean and std.dev: (387.89, 106.81)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 947)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (304, 708, 1146)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2830)\n",
      "[M::mem_pestat] mean and std.dev: (795.57, 646.70)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3672)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (304, 804, 1427)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 3673)\n",
      "[M::mem_pestat] mean and std.dev: (877.33, 758.63)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 4796)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 333400 reads in 291.630 CPU sec, 58.194 real sec\n",
      "[M::process] read 333402 sequences (50000152 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (247, 17839, 2142, 247)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (250, 637, 1437)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 3811)\n",
      "[M::mem_pestat] mean and std.dev: (822.13, 758.35)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 4998)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (346, 410, 497)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (44, 799)\n",
      "[M::mem_pestat] mean and std.dev: (386.21, 107.53)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 950)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (314, 707, 1206)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2990)\n",
      "[M::mem_pestat] mean and std.dev: (819.02, 669.89)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3882)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (372, 807, 1479)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 3693)\n",
      "[M::mem_pestat] mean and std.dev: (897.42, 731.42)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 4800)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 333404 reads in 300.532 CPU sec, 59.984 real sec\n",
      "[M::process] read 333406 sequences (50000209 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (255, 17701, 2074, 235)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (284, 647, 1379)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 3569)\n",
      "[M::mem_pestat] mean and std.dev: (832.72, 752.19)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 4664)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (345, 411, 500)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (35, 810)\n",
      "[M::mem_pestat] mean and std.dev: (385.69, 109.29)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 965)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (306, 718, 1225)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 3063)\n",
      "[M::mem_pestat] mean and std.dev: (855.47, 722.75)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3982)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (312, 650, 1418)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 3630)\n",
      "[M::mem_pestat] mean and std.dev: (830.03, 793.58)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 4736)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 333402 reads in 303.688 CPU sec, 60.612 real sec\n",
      "[M::process] read 268266 sequences (40231683 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (278, 18125, 2173, 230)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (230, 602, 1372)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 3656)\n",
      "[M::mem_pestat] mean and std.dev: (731.33, 669.11)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 4798)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (345, 410, 498)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (39, 804)\n",
      "[M::mem_pestat] mean and std.dev: (385.61, 109.12)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 957)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (291, 687, 1165)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2913)\n",
      "[M::mem_pestat] mean and std.dev: (799.39, 674.67)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3787)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (289, 604, 1395)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 3607)\n",
      "[M::mem_pestat] mean and std.dev: (827.82, 743.66)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 4713)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 333406 reads in 295.321 CPU sec, 58.976 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (224, 14385, 1775, 182)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (214, 537, 1164)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 3064)\n",
      "[M::mem_pestat] mean and std.dev: (697.64, 696.82)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 4014)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (347, 410, 500)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (41, 806)\n",
      "[M::mem_pestat] mean and std.dev: (387.51, 107.43)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 959)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (312, 744, 1207)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2997)\n",
      "[M::mem_pestat] mean and std.dev: (869.08, 725.12)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3892)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (269, 539, 1471)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 3875)\n",
      "[M::mem_pestat] mean and std.dev: (802.32, 753.04)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 5077)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 268266 reads in 242.919 CPU sec, 48.551 real sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa mem -t 5 /home/hqyone/mnt/2tb/eccDNA/genome/hg38_p13/GRCh38.p13.genome.fa /home/hqyone/mnt/2tb/eccDNA/data-raw/GHS21030056/Sample_4-4//for_pair.fq.gz /home/hqyone/mnt/2tb/eccDNA/data-raw/GHS21030056/Sample_4-4//rev_pair.fq.gz\n",
      "[main] Real time: 28758.911 sec; CPU: 143930.788 sec\n",
      "[bam_sort_core] merging from 72 files and 4 in-memory blocks...\n"
     ]
    }
   ],
   "source": [
    "# alignmnemt bwa + samtools\n",
    "bam_out_dir=output_dir+\"/bam\"\n",
    "![ ! -d $bam_out_dir ] && mkdir $bam_out_dir\n",
    "bamfile=bam_out_dir+\"/\"+sample_id+\".bam\"\n",
    "sort_bamfile=bam_out_dir+\"/\"+sample_id+\"_sort.bam\"\n",
    "!bwa mem -t 5 $ref_fa $output_forward_paired_fq_gz $output_reverse_paired_fq_gz> $bamfile\n",
    "!samtools sort $bamfile -o $sort_bamfile -@ 4\n",
    "!samtools index $sort_bamfile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "needed-friday",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "INFO\t2021-07-02 08:57:02\tMarkDuplicates\t\n",
      "\n",
      "********** NOTE: Picard's command line syntax is changing.\n",
      "**********\n",
      "********** For more information, please see:\n",
      "********** https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)\n",
      "**********\n",
      "********** The command line looks like this in the new syntax:\n",
      "**********\n",
      "**********    MarkDuplicates -I /home/hqyone/mnt/2tb/eccDNA/data-raw/GHS21030056/Sample_4-4//bam/Sample_4-4_sort.bam -O /home/hqyone/mnt/2tb/eccDNA/data-raw/GHS21030056/Sample_4-4//bam/Sample_4-4_sort_du.bam -M /home/hqyone/mnt/2tb/eccDNA/data-raw/GHS21030056/Sample_4-4//bam/Sample_4-4_marked_dup_metrics.txt -REMOVE_DUPLICATES True\n",
      "**********\n",
      "\n",
      "\n",
      "08:57:03.279 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/media/hqyone/2tb/eccDNA/code/tools/picard.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Fri Jul 02 08:57:03 MST 2021] MarkDuplicates INPUT=[/home/hqyone/mnt/2tb/eccDNA/data-raw/GHS21030056/Sample_4-4//bam/Sample_4-4_sort.bam] OUTPUT=/home/hqyone/mnt/2tb/eccDNA/data-raw/GHS21030056/Sample_4-4/bam/Sample_4-4_sort_du.bam METRICS_FILE=/home/hqyone/mnt/2tb/eccDNA/data-raw/GHS21030056/Sample_4-4/bam/Sample_4-4_marked_dup_metrics.txt REMOVE_DUPLICATES=true    MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP=50000 MAX_FILE_HANDLES_FOR_READ_ENDS_MAP=8000 SORTING_COLLECTION_SIZE_RATIO=0.25 TAG_DUPLICATE_SET_MEMBERS=false REMOVE_SEQUENCING_DUPLICATES=false TAGGING_POLICY=DontTag CLEAR_DT=true DUPLEX_UMI=false ADD_PG_TAG_TO_READS=true ASSUME_SORTED=false DUPLICATE_SCORING_STRATEGY=SUM_OF_BASE_QUALITIES PROGRAM_RECORD_ID=MarkDuplicates PROGRAM_GROUP_NAME=MarkDuplicates READ_NAME_REGEX=<optimized capture of last three ':' separated fields as numeric values> OPTICAL_DUPLICATE_PIXEL_DISTANCE=100 MAX_OPTICAL_DUPLICATE_SET_SIZE=300000 VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false GA4GH_CLIENT_SECRETS=client_secrets.json USE_JDK_DEFLATER=false USE_JDK_INFLATER=false\n",
      "[Fri Jul 02 08:57:03 MST 2021] Executing as hqyone@hqyone-sev1 on Linux 5.8.0-53-generic amd64; OpenJDK 64-Bit Server VM 11.0.11+9-Ubuntu-0ubuntu2.20.04; Deflater: Intel; Inflater: Intel; Provider GCS is not available; Picard version: 2.25.0\n",
      "INFO\t2021-07-02 08:57:03\tMarkDuplicates\tStart of doWork freeMemory: 136367544; totalMemory: 142606336; maxMemory: 16844324864\n",
      "INFO\t2021-07-02 08:57:03\tMarkDuplicates\tReading input file and constructing read end information.\n",
      "INFO\t2021-07-02 08:57:03\tMarkDuplicates\tWill retain up to 61030162 data points before spilling to disk.\n",
      "INFO\t2021-07-02 08:57:05\tMarkDuplicates\tRead     1,000,000 records.  Elapsed time: 00:00:02s.  Time for last 1,000,000:    2s.  Last read position: chr1:22,506,070\n",
      "INFO\t2021-07-02 08:57:05\tMarkDuplicates\tTracking 88410 as yet unmatched pairs. 80648 records in RAM.\n",
      "INFO\t2021-07-02 08:57:07\tMarkDuplicates\tRead     2,000,000 records.  Elapsed time: 00:00:04s.  Time for last 1,000,000:    2s.  Last read position: chr1:22,511,940\n",
      "INFO\t2021-07-02 08:57:07\tMarkDuplicates\tTracking 123651 as yet unmatched pairs. 115235 records in RAM.\n",
      "INFO\t2021-07-02 08:57:09\tMarkDuplicates\tRead     3,000,000 records.  Elapsed time: 00:00:06s.  Time for last 1,000,000:    1s.  Last read position: chr1:22,524,244\n",
      "INFO\t2021-07-02 08:57:09\tMarkDuplicates\tTracking 105379 as yet unmatched pairs. 96484 records in RAM.\n",
      "INFO\t2021-07-02 08:57:11\tMarkDuplicates\tRead     4,000,000 records.  Elapsed time: 00:00:08s.  Time for last 1,000,000:    2s.  Last read position: chr1:120,431,397\n",
      "INFO\t2021-07-02 08:57:11\tMarkDuplicates\tTracking 46916 as yet unmatched pairs. 1924 records in RAM.\n",
      "INFO\t2021-07-02 08:57:13\tMarkDuplicates\tRead     5,000,000 records.  Elapsed time: 00:00:10s.  Time for last 1,000,000:    1s.  Last read position: chr1:234,427,791\n",
      "INFO\t2021-07-02 08:57:13\tMarkDuplicates\tTracking 95931 as yet unmatched pairs. 35747 records in RAM.\n",
      "INFO\t2021-07-02 08:57:15\tMarkDuplicates\tRead     6,000,000 records.  Elapsed time: 00:00:12s.  Time for last 1,000,000:    2s.  Last read position: chr2:20,182,553\n",
      "INFO\t2021-07-02 08:57:15\tMarkDuplicates\tTracking 70536 as yet unmatched pairs. 7623 records in RAM.\n",
      "INFO\t2021-07-02 08:57:17\tMarkDuplicates\tRead     7,000,000 records.  Elapsed time: 00:00:14s.  Time for last 1,000,000:    2s.  Last read position: chr2:176,208,768\n",
      "INFO\t2021-07-02 08:57:17\tMarkDuplicates\tTracking 101908 as yet unmatched pairs. 13836 records in RAM.\n",
      "INFO\t2021-07-02 08:57:19\tMarkDuplicates\tRead     8,000,000 records.  Elapsed time: 00:00:16s.  Time for last 1,000,000:    2s.  Last read position: chr3:49,328,708\n",
      "INFO\t2021-07-02 08:57:19\tMarkDuplicates\tTracking 114611 as yet unmatched pairs. 11031 records in RAM.\n",
      "INFO\t2021-07-02 08:57:21\tMarkDuplicates\tRead     9,000,000 records.  Elapsed time: 00:00:18s.  Time for last 1,000,000:    1s.  Last read position: chr3:151,391,450\n",
      "INFO\t2021-07-02 08:57:21\tMarkDuplicates\tTracking 111869 as yet unmatched pairs. 1034 records in RAM.\n",
      "INFO\t2021-07-02 08:57:23\tMarkDuplicates\tRead    10,000,000 records.  Elapsed time: 00:00:20s.  Time for last 1,000,000:    1s.  Last read position: chr4:93,464,623\n",
      "INFO\t2021-07-02 08:57:23\tMarkDuplicates\tTracking 122030 as yet unmatched pairs. 1755 records in RAM.\n",
      "INFO\t2021-07-02 08:57:25\tMarkDuplicates\tRead    11,000,000 records.  Elapsed time: 00:00:22s.  Time for last 1,000,000:    2s.  Last read position: chr5:33,132,257\n",
      "INFO\t2021-07-02 08:57:25\tMarkDuplicates\tTracking 129652 as yet unmatched pairs. 7019 records in RAM.\n",
      "INFO\t2021-07-02 08:57:27\tMarkDuplicates\tRead    12,000,000 records.  Elapsed time: 00:00:23s.  Time for last 1,000,000:    1s.  Last read position: chr5:62,908,287\n",
      "INFO\t2021-07-02 08:57:27\tMarkDuplicates\tTracking 157050 as yet unmatched pairs. 10785 records in RAM.\n",
      "INFO\t2021-07-02 08:57:29\tMarkDuplicates\tRead    13,000,000 records.  Elapsed time: 00:00:25s.  Time for last 1,000,000:    1s.  Last read position: chr5:117,806,249\n",
      "INFO\t2021-07-02 08:57:29\tMarkDuplicates\tTracking 176215 as yet unmatched pairs. 25952 records in RAM.\n",
      "INFO\t2021-07-02 08:57:31\tMarkDuplicates\tRead    14,000,000 records.  Elapsed time: 00:00:27s.  Time for last 1,000,000:    2s.  Last read position: chr6:71,654,731\n",
      "INFO\t2021-07-02 08:57:31\tMarkDuplicates\tTracking 175404 as yet unmatched pairs. 6291 records in RAM.\n",
      "INFO\t2021-07-02 08:57:33\tMarkDuplicates\tRead    15,000,000 records.  Elapsed time: 00:00:29s.  Time for last 1,000,000:    2s.  Last read position: chr7:23,253,949\n",
      "INFO\t2021-07-02 08:57:33\tMarkDuplicates\tTracking 203178 as yet unmatched pairs. 31660 records in RAM.\n",
      "INFO\t2021-07-02 08:57:35\tMarkDuplicates\tRead    16,000,000 records.  Elapsed time: 00:00:31s.  Time for last 1,000,000:    1s.  Last read position: chr7:155,898,688\n",
      "INFO\t2021-07-02 08:57:35\tMarkDuplicates\tTracking 206309 as yet unmatched pairs. 21289 records in RAM.\n",
      "INFO\t2021-07-02 08:57:37\tMarkDuplicates\tRead    17,000,000 records.  Elapsed time: 00:00:33s.  Time for last 1,000,000:    2s.  Last read position: chr9:63,438,240\n",
      "INFO\t2021-07-02 08:57:37\tMarkDuplicates\tTracking 196441 as yet unmatched pairs. 11939 records in RAM.\n",
      "INFO\t2021-07-02 08:57:39\tMarkDuplicates\tRead    18,000,000 records.  Elapsed time: 00:00:35s.  Time for last 1,000,000:    1s.  Last read position: chr9:120,990,726\n",
      "INFO\t2021-07-02 08:57:39\tMarkDuplicates\tTracking 241734 as yet unmatched pairs. 50508 records in RAM.\n",
      "INFO\t2021-07-02 08:57:41\tMarkDuplicates\tRead    19,000,000 records.  Elapsed time: 00:00:38s.  Time for last 1,000,000:    2s.  Last read position: chr10:74,935,814\n",
      "INFO\t2021-07-02 08:57:41\tMarkDuplicates\tTracking 195206 as yet unmatched pairs. 31825 records in RAM.\n",
      "INFO\t2021-07-02 08:57:43\tMarkDuplicates\tRead    20,000,000 records.  Elapsed time: 00:00:40s.  Time for last 1,000,000:    2s.  Last read position: chr11:9,939,434\n",
      "INFO\t2021-07-02 08:57:43\tMarkDuplicates\tTracking 232868 as yet unmatched pairs. 36149 records in RAM.\n",
      "INFO\t2021-07-02 08:57:45\tMarkDuplicates\tRead    21,000,000 records.  Elapsed time: 00:00:42s.  Time for last 1,000,000:    1s.  Last read position: chr11:16,869,355\n",
      "INFO\t2021-07-02 08:57:45\tMarkDuplicates\tTracking 276088 as yet unmatched pairs. 77228 records in RAM.\n",
      "INFO\t2021-07-02 08:57:47\tMarkDuplicates\tRead    22,000,000 records.  Elapsed time: 00:00:44s.  Time for last 1,000,000:    2s.  Last read position: chr11:86,250,412\n",
      "INFO\t2021-07-02 08:57:47\tMarkDuplicates\tTracking 206263 as yet unmatched pairs. 2899 records in RAM.\n",
      "INFO\t2021-07-02 08:57:49\tMarkDuplicates\tRead    23,000,000 records.  Elapsed time: 00:00:45s.  Time for last 1,000,000:    1s.  Last read position: chr11:129,869,666\n",
      "INFO\t2021-07-02 08:57:49\tMarkDuplicates\tTracking 307676 as yet unmatched pairs. 102687 records in RAM.\n",
      "INFO\t2021-07-02 08:57:51\tMarkDuplicates\tRead    24,000,000 records.  Elapsed time: 00:00:47s.  Time for last 1,000,000:    1s.  Last read position: chr11:129,874,501\n",
      "INFO\t2021-07-02 08:57:51\tMarkDuplicates\tTracking 218123 as yet unmatched pairs. 12967 records in RAM.\n",
      "INFO\t2021-07-02 08:57:53\tMarkDuplicates\tRead    25,000,000 records.  Elapsed time: 00:00:49s.  Time for last 1,000,000:    2s.  Last read position: chr12:62,603,418\n",
      "INFO\t2021-07-02 08:57:53\tMarkDuplicates\tTracking 203903 as yet unmatched pairs. 11351 records in RAM.\n",
      "INFO\t2021-07-02 08:57:55\tMarkDuplicates\tRead    26,000,000 records.  Elapsed time: 00:00:51s.  Time for last 1,000,000:    2s.  Last read position: chr13:99,073,196\n",
      "INFO\t2021-07-02 08:57:55\tMarkDuplicates\tTracking 197112 as yet unmatched pairs. 2821 records in RAM.\n",
      "INFO\t2021-07-02 08:57:57\tMarkDuplicates\tRead    27,000,000 records.  Elapsed time: 00:00:53s.  Time for last 1,000,000:    2s.  Last read position: chr15:60,346,382\n",
      "INFO\t2021-07-02 08:57:57\tMarkDuplicates\tTracking 191893 as yet unmatched pairs. 7573 records in RAM.\n",
      "INFO\t2021-07-02 08:57:59\tMarkDuplicates\tRead    28,000,000 records.  Elapsed time: 00:00:56s.  Time for last 1,000,000:    2s.  Last read position: chr16:9,481,060\n",
      "INFO\t2021-07-02 08:57:59\tMarkDuplicates\tTracking 189621 as yet unmatched pairs. 30142 records in RAM.\n",
      "INFO\t2021-07-02 08:58:01\tMarkDuplicates\tRead    29,000,000 records.  Elapsed time: 00:00:57s.  Time for last 1,000,000:    1s.  Last read position: chr16:69,357,518\n",
      "INFO\t2021-07-02 08:58:01\tMarkDuplicates\tTracking 332123 as yet unmatched pairs. 161868 records in RAM.\n",
      "INFO\t2021-07-02 08:58:03\tMarkDuplicates\tRead    30,000,000 records.  Elapsed time: 00:00:59s.  Time for last 1,000,000:    1s.  Last read position: chr16:69,367,079\n",
      "INFO\t2021-07-02 08:58:03\tMarkDuplicates\tTracking 354121 as yet unmatched pairs. 183750 records in RAM.\n",
      "INFO\t2021-07-02 08:58:04\tMarkDuplicates\tRead    31,000,000 records.  Elapsed time: 00:01:01s.  Time for last 1,000,000:    1s.  Last read position: chr16:69,384,589\n",
      "INFO\t2021-07-02 08:58:04\tMarkDuplicates\tTracking 288870 as yet unmatched pairs. 118373 records in RAM.\n",
      "INFO\t2021-07-02 08:58:06\tMarkDuplicates\tRead    32,000,000 records.  Elapsed time: 00:01:03s.  Time for last 1,000,000:    2s.  Last read position: chr17:74,031,555\n",
      "INFO\t2021-07-02 08:58:06\tMarkDuplicates\tTracking 174270 as yet unmatched pairs. 8970 records in RAM.\n",
      "INFO\t2021-07-02 08:58:09\tMarkDuplicates\tRead    33,000,000 records.  Elapsed time: 00:01:06s.  Time for last 1,000,000:    2s.  Last read position: chr19:41,916,769\n",
      "INFO\t2021-07-02 08:58:09\tMarkDuplicates\tTracking 117466 as yet unmatched pairs. 7607 records in RAM.\n",
      "INFO\t2021-07-02 08:58:11\tMarkDuplicates\tRead    34,000,000 records.  Elapsed time: 00:01:08s.  Time for last 1,000,000:    2s.  Last read position: chr20:43,360,499\n",
      "INFO\t2021-07-02 08:58:11\tMarkDuplicates\tTracking 135363 as yet unmatched pairs. 25998 records in RAM.\n",
      "INFO\t2021-07-02 08:58:13\tMarkDuplicates\tRead    35,000,000 records.  Elapsed time: 00:01:10s.  Time for last 1,000,000:    2s.  Last read position: chr22:43,524,003\n",
      "INFO\t2021-07-02 08:58:13\tMarkDuplicates\tTracking 105643 as yet unmatched pairs. 1558 records in RAM.\n",
      "INFO\t2021-07-02 08:58:15\tMarkDuplicates\tRead    36,000,000 records.  Elapsed time: 00:01:12s.  Time for last 1,000,000:    2s.  Last read position: chrX:39,916,247\n",
      "INFO\t2021-07-02 08:58:15\tMarkDuplicates\tTracking 279074 as yet unmatched pairs. 197125 records in RAM.\n",
      "INFO\t2021-07-02 08:58:17\tMarkDuplicates\tRead    37,000,000 records.  Elapsed time: 00:01:14s.  Time for last 1,000,000:    2s.  Last read position: chrX:141,374,112\n",
      "INFO\t2021-07-02 08:58:17\tMarkDuplicates\tTracking 86133 as yet unmatched pairs. 3509 records in RAM.\n",
      "INFO\t2021-07-02 08:58:20\tMarkDuplicates\tRead    38,000,000 records.  Elapsed time: 00:01:16s.  Time for last 1,000,000:    2s.  Last read position: GL000251.2:2,239,245\n",
      "INFO\t2021-07-02 08:58:20\tMarkDuplicates\tTracking 61968 as yet unmatched pairs. 14530 records in RAM.\n",
      "INFO\t2021-07-02 08:58:22\tMarkDuplicates\tRead 38668350 records. 0 pairs never matched.\n",
      "INFO\t2021-07-02 08:58:23\tMarkDuplicates\tAfter buildSortedReadEndLists freeMemory: 2731834976; totalMemory: 6773800960; maxMemory: 16844324864\n",
      "INFO\t2021-07-02 08:58:23\tMarkDuplicates\tWill retain up to 526385152 duplicate indices before spilling to disk.\n",
      "INFO\t2021-07-02 08:58:24\tMarkDuplicates\tTraversing read pair information and detecting duplicates.\n",
      "INFO\t2021-07-02 08:58:27\tMarkDuplicates\tTraversing fragment information and detecting duplicates.\n",
      "INFO\t2021-07-02 08:58:29\tMarkDuplicates\tSorting list of duplicate records.\n",
      "INFO\t2021-07-02 08:58:30\tMarkDuplicates\tAfter generateDuplicateIndexes freeMemory: 6765166920; totalMemory: 10989076480; maxMemory: 16844324864\n",
      "INFO\t2021-07-02 08:58:30\tMarkDuplicates\tMarking 13195322 records as duplicates.\n",
      "INFO\t2021-07-02 08:58:30\tMarkDuplicates\tFound 570973 optical duplicate clusters.\n",
      "INFO\t2021-07-02 08:58:30\tMarkDuplicates\tReads are assumed to be ordered by: coordinate\n",
      "INFO\t2021-07-02 08:59:41\tMarkDuplicates\tWritten    10,000,000 records.  Elapsed time: 00:01:11s.  Time for last 10,000,000:   71s.  Last read position: chr7:41,876,913\n",
      "INFO\t2021-07-02 09:00:51\tMarkDuplicates\tWritten    20,000,000 records.  Elapsed time: 00:02:21s.  Time for last 10,000,000:   69s.  Last read position: chr16:69,370,578\n",
      "INFO\t2021-07-02 09:02:13\tMarkDuplicates\tWritten    30,000,000 records.  Elapsed time: 00:03:43s.  Time for last 10,000,000:   82s.  Last read position: */*\n",
      "INFO\t2021-07-02 09:03:46\tMarkDuplicates\tWritten    40,000,000 records.  Elapsed time: 00:05:16s.  Time for last 10,000,000:   93s.  Last read position: */*\n",
      "INFO\t2021-07-02 09:05:19\tMarkDuplicates\tWritten    50,000,000 records.  Elapsed time: 00:06:49s.  Time for last 10,000,000:   92s.  Last read position: */*\n",
      "INFO\t2021-07-02 09:06:53\tMarkDuplicates\tWritten    60,000,000 records.  Elapsed time: 00:08:22s.  Time for last 10,000,000:   93s.  Last read position: */*\n",
      "INFO\t2021-07-02 09:08:26\tMarkDuplicates\tWritten    70,000,000 records.  Elapsed time: 00:09:56s.  Time for last 10,000,000:   93s.  Last read position: */*\n",
      "INFO\t2021-07-02 09:09:58\tMarkDuplicates\tWritten    80,000,000 records.  Elapsed time: 00:11:28s.  Time for last 10,000,000:   92s.  Last read position: */*\n",
      "INFO\t2021-07-02 09:11:31\tMarkDuplicates\tWritten    90,000,000 records.  Elapsed time: 00:13:00s.  Time for last 10,000,000:   92s.  Last read position: */*\n",
      "INFO\t2021-07-02 09:13:08\tMarkDuplicates\tWritten   100,000,000 records.  Elapsed time: 00:14:37s.  Time for last 10,000,000:   97s.  Last read position: */*\n",
      "INFO\t2021-07-02 09:14:44\tMarkDuplicates\tWritten   110,000,000 records.  Elapsed time: 00:16:14s.  Time for last 10,000,000:   96s.  Last read position: */*\n",
      "INFO\t2021-07-02 09:16:20\tMarkDuplicates\tWritten   120,000,000 records.  Elapsed time: 00:17:50s.  Time for last 10,000,000:   95s.  Last read position: */*\n",
      "INFO\t2021-07-02 09:17:56\tMarkDuplicates\tWritten   130,000,000 records.  Elapsed time: 00:19:26s.  Time for last 10,000,000:   96s.  Last read position: */*\n",
      "INFO\t2021-07-02 09:19:32\tMarkDuplicates\tWritten   140,000,000 records.  Elapsed time: 00:21:02s.  Time for last 10,000,000:   96s.  Last read position: */*\n",
      "INFO\t2021-07-02 09:21:09\tMarkDuplicates\tWritten   150,000,000 records.  Elapsed time: 00:22:39s.  Time for last 10,000,000:   96s.  Last read position: */*\n",
      "INFO\t2021-07-02 09:21:25\tMarkDuplicates\tWriting complete. Closing input iterator.\n",
      "INFO\t2021-07-02 09:21:25\tMarkDuplicates\tDuplicate Index cleanup.\n",
      "INFO\t2021-07-02 09:21:25\tMarkDuplicates\tGetting Memory Stats.\n",
      "INFO\t2021-07-02 09:21:26\tMarkDuplicates\tBefore output close freeMemory: 117443304; totalMemory: 125829120; maxMemory: 16844324864\n",
      "INFO\t2021-07-02 09:21:26\tMarkDuplicates\tClosed outputs. Getting more Memory Stats.\n",
      "INFO\t2021-07-02 09:21:26\tMarkDuplicates\tAfter output close freeMemory: 51172016; totalMemory: 58720256; maxMemory: 16844324864\n",
      "[Fri Jul 02 09:21:26 MST 2021] picard.sam.markduplicates.MarkDuplicates done. Elapsed time: 24.38 minutes.\n",
      "Runtime.totalMemory()=58720256\n"
     ]
    }
   ],
   "source": [
    "# remove duplication by picard tools\n",
    "# download form https://github.com/broadinstitute/picard/releases/download/2.25.6/picard.jar\n",
    "picard=\"/home/hqyone/mnt/2tb/eccDNA/code/tools/picard.jar\"\n",
    "bam_out_dir=output_dir+\"/bam\"\n",
    "sort_bamfile=bam_out_dir+\"/\"+sample_id+\"_sort.bam\"\n",
    "sort_du_bamfile=bam_out_dir+\"/\"+sample_id+\"_sort_du.bam\"\n",
    "marked_dup_metrix= bam_out_dir+\"/\"+sample_id+\"_marked_dup_metrics.txt\"\n",
    "!java -jar $picard MarkDuplicates I=$sort_bamfile O=$sort_du_bamfile M=$marked_dup_metrix REMOVE_DUPLICATES=True "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "shaped-monitor",
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "Error: Unable to access jarfile /home/hqyone/mnt/2t/eccDNA/code/tools/picard.jar\n"
     ]
    }
   ],
   "source": [
    "# Generate bam file Sorted with queryname\n",
    "picard=\"/home/hqyone/mnt/2t/eccDNA/code/tools/picard.jar\"\n",
    "bam_out_dir=output_dir+\"/bam\"\n",
    "sort_du_bamfile=bam_out_dir+\"/\"+sample_id+\"_sort_du.bam\"\n",
    "name_sort_bamfile=bam_out_dir+\"/\"+sample_id+\"_ns_sort_du.bam\"\n",
    "!java -jar $picard SortSam I=$sort_du_bamfile O=$name_sort_bamfile SORT_ORDER=queryname\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "permanent-boxing",
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "INFO  @ Fri, 02 Jul 2021 09:21:27: \n",
      "# Command line: callpeak -t /home/hqyone/mnt/2tb/eccDNA/data-raw/GHS21030056/Sample_4-4//bam/Sample_4-4_sort_du.bam --broad -g hs --broad-cutoff 0.1 -B -f BAMPE -np --outdir /home/hqyone/mnt/2tb/eccDNA/data-raw/GHS21030056/Sample_4-4//macs2\n",
      "# ARGUMENTS LIST:\n",
      "# name = p\n",
      "# format = BAMPE\n",
      "# ChIP-seq file = ['/home/hqyone/mnt/2tb/eccDNA/data-raw/GHS21030056/Sample_4-4//bam/Sample_4-4_sort_du.bam']\n",
      "# control file = None\n",
      "# effective genome size = 2.70e+09\n",
      "# band width = 300\n",
      "# model fold = [5, 50]\n",
      "# qvalue cutoff for narrow/strong regions = 5.00e-02\n",
      "# qvalue cutoff for broad/weak regions = 1.00e-01\n",
      "# The maximum gap between significant sites is assigned as the read length/tag size.\n",
      "# The minimum length of peaks is assigned as the predicted fragment length \"d\".\n",
      "# Larger dataset will be scaled towards smaller dataset.\n",
      "# Range for calculating regional lambda is: 10000 bps\n",
      "# Broad region calling is on\n",
      "# Paired-End mode is on\n",
      " \n",
      "INFO  @ Fri, 02 Jul 2021 09:21:27: #1 read fragment files... \n",
      "INFO  @ Fri, 02 Jul 2021 09:21:27: #1 read treatment fragments... \n",
      "INFO  @ Fri, 02 Jul 2021 09:21:38:  1000000 \n",
      "INFO  @ Fri, 02 Jul 2021 09:21:43:  2000000 \n",
      "INFO  @ Fri, 02 Jul 2021 09:21:49:  3000000 \n",
      "INFO  @ Fri, 02 Jul 2021 09:21:54:  4000000 \n",
      "INFO  @ Fri, 02 Jul 2021 09:21:59:  5000000 \n",
      "INFO  @ Fri, 02 Jul 2021 09:22:06:  6000000 \n",
      "INFO  @ Fri, 02 Jul 2021 09:22:14:  7000000 \n",
      "INFO  @ Fri, 02 Jul 2021 09:22:19:  8000000 \n",
      "INFO  @ Fri, 02 Jul 2021 09:25:51: 8615323 fragments have been read. \n",
      "INFO  @ Fri, 02 Jul 2021 09:25:54: #1 mean fragment size is determined as 385.8 bp from treatment \n",
      "INFO  @ Fri, 02 Jul 2021 09:25:54: #1 fragment size = 385.8 \n",
      "INFO  @ Fri, 02 Jul 2021 09:25:54: #1  total fragments in treatment: 8615323 \n",
      "INFO  @ Fri, 02 Jul 2021 09:25:54: #1 user defined the maximum fragments... \n",
      "INFO  @ Fri, 02 Jul 2021 09:25:54: #1 filter out redundant fragments by allowing at most 1 identical fragment(s) \n",
      "INFO  @ Fri, 02 Jul 2021 09:26:03: #1  fragments after filtering in treatment: 6241944 \n",
      "INFO  @ Fri, 02 Jul 2021 09:26:03: #1  Redundant rate of treatment: 0.28 \n",
      "INFO  @ Fri, 02 Jul 2021 09:26:03: #1 finished! \n",
      "INFO  @ Fri, 02 Jul 2021 09:26:03: #2 Build Peak Model... \n",
      "INFO  @ Fri, 02 Jul 2021 09:26:03: #2 Skipped... \n",
      "INFO  @ Fri, 02 Jul 2021 09:26:03: #3 Call peaks... \n",
      "INFO  @ Fri, 02 Jul 2021 09:26:03: #3 Call broad peaks with given level1 -log10qvalue cutoff and level2: 1.301030, 1.000000... \n",
      "INFO  @ Fri, 02 Jul 2021 09:26:03: #3 Pre-compute pvalue-qvalue table... \n",
      "INFO  @ Fri, 02 Jul 2021 09:26:56: #3 In the peak calling step, the following will be performed simultaneously: \n",
      "INFO  @ Fri, 02 Jul 2021 09:26:56: #3   Write bedGraph files for treatment pileup (after scaling if necessary)... p_treat_pileup.bdg \n",
      "INFO  @ Fri, 02 Jul 2021 09:26:56: #3   Write bedGraph files for control lambda (after scaling if necessary)... p_control_lambda.bdg \n",
      "INFO  @ Fri, 02 Jul 2021 09:26:56: #3 Call peaks for each chromosome... \n",
      "INFO  @ Fri, 02 Jul 2021 09:27:05: #4 Write output xls file... /home/hqyone/mnt/2tb/eccDNA/data-raw/GHS21030056/Sample_4-4//macs2/p_peaks.xls \n",
      "INFO  @ Fri, 02 Jul 2021 09:27:05: #4 Write broad peak in broadPeak format file... /home/hqyone/mnt/2tb/eccDNA/data-raw/GHS21030056/Sample_4-4//macs2/p_peaks.broadPeak \n",
      "INFO  @ Fri, 02 Jul 2021 09:27:05: #4 Write broad peak in bed12/gappedPeak format file... /home/hqyone/mnt/2tb/eccDNA/data-raw/GHS21030056/Sample_4-4//macs2/p_peaks.gappedPeak \n",
      "INFO  @ Fri, 02 Jul 2021 09:27:05: Done! \n"
     ]
    }
   ],
   "source": [
    "# identified regions using MACS\n",
    "macs_out_dir=output_dir+\"/macs2\"\n",
    "bam_out_dir=output_dir+\"/bam\"\n",
    "sort_du_bamfile=bam_out_dir+\"/\"+sample_id+\"_sort_du.bam\"\n",
    "![ ! -d $macs_out_dir ] && mkdir $macs_out_dir\n",
    "!macs2 callpeak -t $sort_du_bamfile --broad -g hs --broad-cutoff 0.1 -B -f BAMPE -np --outdir $macs_out_dir\n"
   ]
  },
  {
   "source": [
    "# generate bigwig file\n",
    "macs_out_dir=output_dir+\"/macs2\"\n",
    "F=macs_out_dir+\"/p_treat_pileup.bdg\"\n",
    "G=\"/home/hqyone/mnt/2tb/eccDNA/genome/chrom_size/hg38.chrom.sizes\"\n",
    "bedClip=\"/home/hqyone/mnt/2tb/eccDNA/code/tools/bedClip\"\n",
    "bedGraphToBigWig = \"/home/hqyone/mnt/2tb/eccDNA/code/tools/bedGraphToBigWig\"\n",
    "bedtools=\"/home/hqyone/anaconda3/envs/eccDNA/bin/bedtools\"\n",
    "\n",
    "!$bedtools slop -i $F -g $G -b 0 | $bedClip stdin $G {F}_clip\n",
    "!LC_COLLATE=C sort -k1,1 -k2,2n {F}_clip > {F}_sort_clip\n",
    "!$bedGraphToBigWig {F}_sort_clip $G {macs_out_dir}/{sample_id}.bigwig\n",
    "!rm -f {F}_clip {F}_sort_clip"
   ],
   "cell_type": "code",
   "metadata": {},
   "execution_count": 19,
   "outputs": []
  },
  {
   "source": [
    "# annotate regions\n",
    "bedClip=\"/home/hqyone/mnt/2t/eccDNA/code/tools/bedClip\"\n",
    "bedtools=\"/home/hqyone/anaconda3/envs/eccDNA/bin/bedtools\"\n",
    "!{bedtools} slop\n",
    "!{bedClip} \n",
    "!bedtools"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "municipal-niagara",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "\n",
      "\tPeak file = /home/hqyone/mnt/2tb/eccDNA/data-raw/GHS21030056/Sample_4-4//macs2/p_peaks.bed\n",
      "\tGenome = hg38\n",
      "\tOrganism = human\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 4998\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\t\tDuplicated Peak IDs: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 4998\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tReading Positions...\n",
      "\t-----------------------\n",
      "\tFinding Closest TSS...\n",
      "\tAnnotating:.........................\n",
      "\t\tAnnotation\tNumber of peaks\tTotal size (bp)\tLog2 Ratio (obs/exp)\tLogP enrichment (+values depleted)\n",
      "\t\t3UTR\t17.0\t26833139\t-1.354\t12.436\n",
      "\t\tmiRNA\t0.0\t97618\t-0.212\t0.158\n",
      "\t\tncRNA\t8.0\t7044597\t-0.512\t1.622\n",
      "\t\tTTS\t32.0\t32407530\t-0.714\t6.469\n",
      "\t\tpseudo\t6.0\t2111155\t0.812\t-2.027\n",
      "\t\tExon\t40.0\t37120946\t-0.588\t5.610\n",
      "\t\tIntron\t1134.0\t1257910936\t-0.845\t368.020\n",
      "\t\tIntergenic\t3720.0\t1684360386\t0.448\t-423.775\n",
      "\t\tPromoter\t32.0\t35948263\t-0.863\t9.033\n",
      "\t\t5UTR\t2.0\t2601483\t-1.075\t1.568\n",
      "\t\tsnoRNA\t0.0\t357\t-0.001\t0.001\n",
      "\t\tscRNA\t0.0\t97\t-0.000\t0.000\n",
      "\t\trRNA\t7.0\t25562\t7.402\t-30.858\n",
      "\tNOTE: If this part takes more than 2 minutes, there is a good chance\n",
      "\t\tyour machine ran out of memory: consider hitting ctrl+C and rerunning\n",
      "\t\tthe command with \"-noann\"\n",
      "\tTo capture annotation stats in a file, use \"-annStats <filename>\" next time\n",
      "\tAnnotating:.........................\n",
      "\t\tAnnotation\tNumber of peaks\tTotal size (bp)\tLog2 Ratio (obs/exp)\tLogP enrichment (+values depleted)\n",
      "\t\t3UTR\t17.0\t26833139\t-1.353\t12.422\n",
      "\t\tRetroposon\t4.0\t4062585\t-0.717\t1.536\n",
      "\t\tRC?\t0.0\t59773\t-0.133\t0.097\n",
      "\t\tRNA\t0.0\t108000\t-0.232\t0.175\n",
      "\t\tmiRNA\t0.0\t97618\t-0.212\t0.158\n",
      "\t\tncRNA\t8.0\t7044597\t-0.511\t1.620\n",
      "\t\tTTS\t32.0\t32407530\t-0.713\t6.457\n",
      "\t\tLINE\t556.0\t633178497\t-0.882\t158.801\n",
      "\t\tsrpRNA\t0.0\t262214\t-0.510\t0.424\n",
      "\t\tSINE\t446.0\t379673449\t-0.462\t31.495\n",
      "\t\tRC\t1.0\t362795\t0.768\t-0.812\n",
      "\t\ttRNA\t0.0\t91810\t-0.200\t0.149\n",
      "\t\tDNA?\t0.0\t423197\t-0.753\t0.685\n",
      "\t\tpseudo\t6.0\t2111155\t0.812\t-2.029\n",
      "\t\tDNA\t94.0\t99395597\t-0.775\t19.166\n",
      "\t\tExon\t40.0\t37120946\t-0.587\t5.599\n",
      "\t\tIntron\t516.0\t656103354\t-1.041\t213.150\n",
      "\t\tIntergenic\t522.0\t780877963\t-1.276\t353.085\n",
      "\t\tPromoter\t32.0\t35948263\t-0.863\t9.019\n",
      "\t\t5UTR\t2.0\t2601483\t-1.074\t1.567\n",
      "\t\tsnoRNA\t0.0\t357\t-0.001\t0.001\n",
      "\t\tLTR?\t1.0\t1140409\t-0.884\t0.800\n",
      "\t\tscRNA\t1.0\t121638\t2.345\t-1.722\n",
      "\t\tCpG-Island\t11.0\t9000269\t-0.405\t1.538\n",
      "\t\tLow_complexity\t7.0\t5480212\t-0.342\t1.081\n",
      "\t\tLTR\t238.0\t262167686\t-0.834\t55.311\n",
      "\t\tSimple_repeat\t79.0\t34754582\t0.490\t-6.082\n",
      "\t\tsnRNA\t1.0\t314325\t0.975\t-0.919\n",
      "\t\tUnknown\t1.0\t714206\t-0.209\t0.388\n",
      "\t\tSINE?\t0.0\t2674\t-0.006\t0.004\n",
      "\t\tSatellite\t2373.0\t75387841\t4.282\t-5421.457\n",
      "\t\trRNA\t10.0\t202619\t4.930\t-26.560\n",
      "\tCounting Tags in Peaks from each directory...\n",
      "\tOrganism: human\n",
      "\tLoading Gene Informaiton...\n",
      "\tOutputing Annotation File...\n",
      "\tDone annotating peaks file\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# using homer to annotation the peak\n",
    "# install homer: \n",
    "# wget http://homer.ucsd.edu/homer/configureHomer.pl\n",
    "# perl configureHomer.pl install\n",
    "# perl configureHomer.pl -install hg38\n",
    "# annotatePeaks.pl ERpeaks.txt hg18   > outputfile.txt\n",
    "macs_out_dir=output_dir+\"/macs2\"\n",
    "peak_file = macs_out_dir+\"/p_peaks.xls\"\n",
    "peak_bed = macs_out_dir+\"/p_peaks.bed\"\n",
    "anno_peak_bed = macs_out_dir+\"/p_peaks_anno.bed\"\n",
    "\n",
    "with open(peak_file,'r') as INPUT, open(peak_bed,'w') as OUTPUT:\n",
    "    for line in INPUT:\n",
    "        if (line.startswith(\"#\") or not line.startswith(\"chr\")):\n",
    "            continue\n",
    "        c = line.strip().split(\"\\t\")\n",
    "        if len(c)>6:\n",
    "            ls = [c[0],c[1],c[2],c[0]+\":\"+c[1]+\"-\"+c[2],c[5],\"+\"]\n",
    "            OUTPUT.write(\"\\t\".join(ls)+\"\\n\")\n",
    "\n",
    "import os\n",
    "os.environ['PATH'] += os.pathsep + \"/home/hqyone/mnt/2tb/eccDNA/code/tools/homer/bin\"\n",
    "annot_script=\"annotatePeaks.pl\"\n",
    "!$annot_script {peak_bed} hg38 > {anno_peak_bed}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "[bam_sort_core] merging from 11 files and 1 in-memory blocks...\n"
     ]
    }
   ],
   "source": [
    "# Get 1) paired reads and 2) sort with read id\n",
    "# The final file will be used for eccSearch.ipynb\n",
    "sort_du_bamfile=bam_out_dir+\"/\"+sample_id+\"_sort_du.bam\"\n",
    "map_map_bamfile=bam_out_dir+\"/\"+sample_id+\"_map_map.bam\"\n",
    "sort_du_lsort_bamfile=bam_out_dir+\"/\"+sample_id+\"_lsort_du.bam\"\n",
    "!samtools view -u -f 1 -F 12 {sort_du_bamfile}> {map_map_bamfile}\n",
    "!samtools sort -n {map_map_bamfile} -o {sort_du_lsort_bamfile}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "name": "python3",
   "display_name": "Python 3.7.10 64-bit"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  },
  "interpreter": {
   "hash": "e5030792b3492f6b12d94f1f48beca3d8e59ec05fd59d0aaaa48e684281ed297"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}