{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "complex-denial",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: 'defaults' already in 'channels' list, moving to the top\n",
      "Warning: 'bioconda' already in 'channels' list, moving to the top\n",
      "Warning: 'conda-forge' already in 'channels' list, moving to the top\n",
      "\n",
      "CommandNotFoundError: Your shell has not been properly configured to use 'conda activate'.\n",
      "To initialize your shell, run\n",
      "\n",
      "    $ conda init <SHELL_NAME>\n",
      "\n",
      "Currently supported shells are:\n",
      "  - bash\n",
      "  - fish\n",
      "  - tcsh\n",
      "  - xonsh\n",
      "  - zsh\n",
      "  - powershell\n",
      "\n",
      "See 'conda init --help' for more information and options.\n",
      "\n",
      "IMPORTANT: You may need to close and restart your shell after running 'conda init'.\n",
      "\n",
      "\n",
      "Collecting package metadata (current_repodata.json): done\n",
      "Solving environment: failed with initial frozen solve. Retrying with flexible solve.\n",
      "Collecting package metadata (repodata.json): done\n",
      "Solving environment: | ^C\n",
      "failed with initial frozen solve. Retrying with flexible solve.\n",
      "\n",
      "CondaError: KeyboardInterrupt\n",
      "\n",
      "Collecting package metadata (current_repodata.json): \\ ^C\n",
      "failed\n",
      "\n",
      "CondaError: KeyboardInterrupt\n",
      "\n",
      "Collecting package metadata (current_repodata.json): / ^C\n",
      "failed\n",
      "\n",
      "CondaError: KeyboardInterrupt\n",
      "\n",
      "^C\n",
      "\n",
      "CondaError: KeyboardInterrupt\n",
      "\n",
      "\n",
      "CommandNotFoundError: Your shell has not been properly configured to use 'conda deactivate'.\n",
      "To initialize your shell, run\n",
      "\n",
      "    $ conda init <SHELL_NAME>\n",
      "\n",
      "Currently supported shells are:\n",
      "  - bash\n",
      "  - fish\n",
      "  - tcsh\n",
      "  - xonsh\n",
      "  - zsh\n",
      "  - powershell\n",
      "\n",
      "See 'conda init --help' for more information and options.\n",
      "\n",
      "IMPORTANT: You may need to close and restart your shell after running 'conda init'.\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# install software need for the pipeline\n",
    "!conda config --add channels defaults\n",
    "!conda config --add channels bioconda\n",
    "!conda config --add channels conda-forge\n",
    "\n",
    "def createEnv(env_name):\n",
    "    #!conda create -n $env_name python=3.7 anaconda -y\n",
    "    !conda activate $env_name\n",
    "\n",
    "    !conda install -n $env_name macs2==2.1.2 bcftools==1.9 bedtools==2.28.0 bzip2==1.0.6 bwa==0.7.17 bx-python==0.8.2 fastp==0.20.0 fastqc==0.11.8 -y\n",
    "    !conda install -n $env_name git==2.21.0 gmap==2018.07.04 htslib==1.9 libdeflate==1.0 numpy==1.16.2 pysam==0.15.2 rseqc==2.6.4 -y\n",
    "    !conda install -n $env_name samtools==1.9 scipy==1.2.1 star==2.7.0f wget==1.20.1 trimmomatic -y\n",
    "    !conda install -n $env_name pandas matplotlib seabean nxviz -y\n",
    "\n",
    "    !conda deactivate\n",
    "    \n",
    "    \n",
    "\n",
    "createEnv(\"eccDNA\")\n",
    "\n",
    "conda install macs2==2.1.2 bcftools==1.9 bedtools==2.28.0 bzip2==1.0.6 bwa==0.7.17 bx-python==0.8.2 fastp==0.20.0 fastqc==0.11.8 -y\n",
    "#!conda activate eccDNA\n",
    "#!conda install -n eccDNA trimmomatic pandas matplotlib seabean nxviz -y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "disturbed-string",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'gnome_name' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-5-20bb55be01fe>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     23\u001b[0m                 \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msystem\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"unzip -d {parent_dir}/{gnome_name}/gtf/gencode.v37.chr_patch_hapl_scaff.annotation.gtf.gz\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     24\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 25\u001b[0;31m \u001b[0mbuildGenome\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"/home/hqyone/mnt/2t/eccDNA/genome\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"hg38_p13\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-5-20bb55be01fe>\u001b[0m in \u001b[0;36mbuildGenome\u001b[0;34m(parent_dir, genome_name)\u001b[0m\n\u001b[1;32m     10\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0misdir\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mparent_dir\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;32mnot\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0misdir\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mparent_dir\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m\"/\"\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mgenome_name\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 12\u001b[0;31m             \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msystem\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"mkdir {parent_dir}/{gnome_name}\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     13\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;32mnot\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0misfile\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mparent_dir\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m\"/\"\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mgenome_name\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m\"/GRCh38.p13.genome.fa\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     14\u001b[0m             \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msystem\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"wget {REF_FA_URL} -P {parent_dir}/{gnome_name}\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'gnome_name' is not defined"
     ]
    }
   ],
   "source": [
    "# Download reference geneome \n",
    "# Genome sequences/annoation were download form here\n",
    "# https://www.gencodegenes.org/human/\n",
    "import os\n",
    "def buildGenome(parent_dir, genome_name):\n",
    "    REF_FA_URL=\"ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_37/GRCh38.p13.genome.fa.gz\"\n",
    "    REF_GTF_URL=\"ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_37/gencode.v37.chr_patch_hapl_scaff.annotation.gtf.gz\"\n",
    "    if (os.path.isdir(parent_dir)):\n",
    "        if (not os.path.isdir(parent_dir+\"/\"+genome_name)):\n",
    "            os.system(f\"mkdir {parent_dir}/{genome_name}\")\n",
    "        if (not os.path.isfile(parent_dir+\"/\"+genome_name+\"/GRCh38.p13.genome.fa\")):\n",
    "            os.system(f\"wget {REF_FA_URL} -P {parent_dir}/{genome_name}\")\n",
    "            os.system(f\"unzip -d {parent_dir}/{genome_name}/GRCh38.p13.genome.fa.gz\")\n",
    "            os.system(f\"samtools index {parent_dir}/{genome_name}/GRCh38.p13.genome.fa -@ 3\")\n",
    "            os.system(f\"bwa index {parent_dir}/{genome_name}/GRCh38.p13.genome.fa\")\n",
    "        if (not os.path.isfile(parent_dir+\"/\"+genome_name+\"/gencode.v37.chr_patch_hapl_scaff.annotation.gtf\")):\n",
    "            if (not os.path.isdir(parent_dir+\"/\"+genome_name+\"/gft\")):\n",
    "                os.system(f\"mkdir {parent_dir}/{genome_name}/gtf\")\n",
    "            if (not os.path.isfile(parent_dir+\"/\"+genome_name+\"/gft/gencode.v37.chr_patch_hapl_scaff.annotation.gtf\")):\n",
    "                os.system(f\"wget {REF_GTF_URL} -P {parent_dir}/{genome_name}/gtf\")\n",
    "                os.system(f\"unzip -d {parent_dir}/{genome_name}/gtf/gencode.v37.chr_patch_hapl_scaff.annotation.gtf.gz\")\n",
    "        \n",
    "buildGenome(\"/home/hqyone/mnt/2t/eccDNA/genome\",\"hg38_p13\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "sharing-berry",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python3.7",
   "language": "python",
   "name": "python3.7"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
